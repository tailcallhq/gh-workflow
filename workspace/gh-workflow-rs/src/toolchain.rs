use derive_setters::Setters;

use crate::schema::*;

pub enum Version {
    Stable,
    Beta,
    Nightly,
}

#[derive(Setters)]
pub struct RustToolchain {
    version: Version,
    rust_fmt: bool,
    rust_clippy: bool,
    // TODO: add more rust tool chain components
}

impl RustToolchain {
    pub fn to_job(&self) -> Job {
        Job {
            name: Some(format!("Setup Rust Toolchain")),
            runs_on: vec![Runner::default()],
            steps: vec![

                // TODO: expose as typed methods
                Step::default().uses("actions/checkout@v2".to_string()),
                Step::default().uses("actions-rs/toolchain@v1".to_string()),

                //
            ],
            ..Default::default()
        }
    }
}
