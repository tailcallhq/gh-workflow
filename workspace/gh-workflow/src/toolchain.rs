use derive_setters::Setters;

use crate::workflow::*;

///
/// A type-safe representation of the Rust toolchain.
/// Instead of writing the github action for Rust by hand, we can use this
/// struct to generate the github action.
#[derive(Default)]
pub enum Version {
    #[default]
    Stable,
    Beta,
    Nightly,
}
impl Version {
    pub fn to_string(&self) -> String {
        match self {
            Version::Stable => "stable".to_string(),
            Version::Beta => "beta".to_string(),
            Version::Nightly => "nightly".to_string(),
        }
    }
}

#[derive(Setters, Default)]
pub struct RustToolchain {
    version: Version,
    fmt: bool,
    clippy: bool,
    // TODO: add more rust tool chain components
}

impl RustToolchain {
    pub fn to_job(&self) -> Job {
        todo!();
    }
}
