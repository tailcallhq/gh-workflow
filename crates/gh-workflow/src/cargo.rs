use derive_setters::Setters;

use crate::toolchain::Version;
use crate::{Env, Run, Step};

#[derive(Clone, Setters)]
#[setters(strip_option, into)]
pub struct Cargo {
    /// The command to be executed for eg: fmt, clippy, build, test, etc.
    pub command: String,

    /// The unique identifier of the Step.
    pub id: Option<String>,

    /// Name of the Step
    pub name: Option<String>,

    /// Toolchain to be used for example `+nightly`.
    pub toolchain: Option<Version>,

    /// Arguments to be passed to the cargo command.
    #[setters(skip)]
    pub args: Vec<String>,

    pub env: Option<Env>,
}

impl Cargo {
    /// Creates a new `Cargo` instance with the specified command.
    pub fn new<T: ToString>(cmd: T) -> Cargo {
        Cargo {
            command: cmd.to_string(),
            id: Default::default(),
            name: Default::default(),
            toolchain: Default::default(),
            args: Default::default(),
            env: Default::default(),
        }
    }

    pub fn add_env<T: Into<Env>>(mut self, new_env: T) -> Self {
        let mut env = self.env.unwrap_or_default();
        env.0.extend(new_env.into().0);

        self.env = Some(env);
        self
    }

    /// Sets the toolchain to nightly.
    pub fn nightly(mut self) -> Self {
        self.toolchain = Some(Version::Nightly);
        self
    }

    /// Sets the arguments for the cargo command. If arguments are already set,
    /// it will be overwritten.
    pub fn args<T: ToString>(mut self, args: T) -> Self {
        self.args = vec![args.to_string()];
        self
    }

    /// Adds additional arguments to the cargo command.
    pub fn add_args<T: ToString>(mut self, args: T) -> Self {
        self.args.extend(
            args.to_string()
                .split_whitespace()
                .map(|s| s.to_string())
                .collect::<Vec<_>>(),
        );
        self
    }

    /// Adds the arguments to the cargo command when a condition is met.
    pub fn add_args_when<T: ToString>(self, when: bool, args: T) -> Self {
        if when {
            self.add_args(args)
        } else {
            self
        }
    }
}

impl From<Cargo> for Step<Run> {
    fn from(value: Cargo) -> Self {
        let mut command = vec!["cargo".to_string()];

        if let Some(toolchain) = value.toolchain {
            command.push(format!("+{}", toolchain));
        }

        command.push(value.command);

        // Extend the command with non-empty arguments
        command.extend(
            value
                .args
                .into_iter()
                .map(|arg| arg.trim().to_string())
                .filter(|arg| !arg.is_empty()),
        );

        let mut step = Step::run(command.join(" "));

        if let Some(id) = value.id {
            step = step.id(id);
        }

        if let Some(name) = value.name {
            step = step.name(name);
        }

        if let Some(env) = value.env {
            step = step.env(env);
        }

        step
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_cargo() {
        let cargo = Cargo::new("fmt")
            .add_args("--all")
            .add_args("--check")
            .nightly()
            .add_env(Env::github());

        let step: Step<Run> = cargo.into();

        insta::assert_snapshot!(serde_yaml::to_string(&step).unwrap());
    }
}
