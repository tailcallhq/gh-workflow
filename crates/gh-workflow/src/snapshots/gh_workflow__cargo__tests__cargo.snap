---
source: crates/gh-workflow/src/cargo.rs
expression: "serde_yaml::to_string(&step).unwrap()"
---
run: cargo +nightly fmt --all --check
env:
  GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
